editoast/src/projects.rs

```diff
@@ -0,0 +1,186 @@
+use editoast_derive::EditoastError;
+use serde_json::{json, Map, Value};
+use thiserror::Error;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "projects", context = "Self::context")]
+pub enum ProjectError {
+    /// Couldn't found the infra with the given id
+    #[error("Project '{0}', could not be found")]
+    #[editoast_error(status = 404)]
+    NotFound(i64),
+}
+
+impl ProjectError {
+    fn context(&self) -> Map<String, Value> {
+        match self {
+            ProjectError::NotFound(project_id) => json!({
+                "project_id": project_id,
+            })
+            .as_object()
+            .cloned()
+            .unwrap(),
+        }
+    }
+}
+
+#[derive(
+    Clone, QueryableByName, Queryable, Debug, Serialize, Deserialize, AsChangeset, Identifiable,
+)]
+#[diesel(table_name = osrd_infra_project)]
+pub struct Project {
+    pub id: i64,
+    pub name: String,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub creation_date: NaiveDateTime,
+    pub last_modification: NaiveDateTime,
+    pub tags: Option<Vec<String>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProjectData {
+    pub name: String,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub tags: Option<Vec<String>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProjectPatch {
+    pub name: Option<String>,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub tags: Option<Vec<String>>,
+}
+
+impl Project {
+    pub fn create(data: ProjectData, conn: &mut PgConnection) -> Result<Project> {
+        match sql_query(
+        "INSERT INTO osrd_infra_project (name, description, objectives, funders, budget, image, creation_date,last_modification, tags)
+         VALUES ($1, $2, $3, $4, $5, NULL,CURRENT_TIMESTAMP, CURRENT_TIMESTAMP,$6)
+         RETURNING *",
+    )
+    .bind::<Text, _>(data.name)
+    .bind::<Nullable<Text>,_>(data.description)
+    .bind::<Nullable<Text>,_>(data.objectives)
+    .bind::<Nullable<Array<Text>>,_>(data.funders)
+    .bind::<Nullable<Integer>,_>(data.budget)
+    .bind::<Nullable<Array<Text>>,_>(data.tags)
+    .get_result::<Project>(conn)
+    {
+        Ok(project) => Ok(project),
+        Err(err) => Err(err.into()),
+    }
+    }
+
+    pub fn list(conn: &mut PgConnection, page: i64, page_size: i64) -> Vec<Project> {
+        let offset = page_size * (page - 1);
+        dsl::osrd_infra_project
+            .limit(page_size)
+            .offset(offset)
+            .load(conn)
+            .expect("List project query failed")
```

Comment:
You might want to adapt and use the `Pagination::paginate` function. Currently, this function takes raw queries only. We should be able also to accept diesel queries.

This function will wrap the query to retrieve the total number of elements (practical to build the `PaginatedResponse`)

Note: The tooling to handle paginated queries is not perfect and can be improved. Be free to suggest new interfaces.